logger/dapr_logger.go

```diff
@@ -112,65 +123,143 @@ func (l *daprLogger) SetOutput(dst io.Writer) {
+	l.printf(context.Background(), logrus.DebugLevel, format, args...)
+}
+
+// DebugWithContext logs a message and context (traceid...) at level Debug.
+func (l *daprLogger) DebugWithContext(ctx context.Context, args ...interface{}) {
+	l.print(ctx, logrus.DebugLevel, args...)
+}
+
+// DebugWithContext logs a message and context (traceid...) at level Debug.
+func (l *daprLogger) DebugWithContextf(ctx context.Context, format string, args ...interface{}) {
+	l.printf(ctx, logrus.DebugLevel, format, args...)
 }
 
 // Warn logs a message at level Warn.
 func (l *daprLogger) Warn(args ...interface{}) {
-	l.logger.Log(logrus.WarnLevel, args...)
+	l.print(context.Background(), logrus.WarnLevel, args...)
 }
 
 // Warnf logs a message at level Warn.
 func (l *daprLogger) Warnf(format string, args ...interface{}) {
-	l.logger.Logf(logrus.WarnLevel, format, args...)
+	l.printf(context.Background(), logrus.WarnLevel, format, args...)
+}
+
+// WarnWithContext logs a message and context (tarceid...) at level Warn.
+func (l *daprLogger) WarnWithContext(ctx context.Context, args ...interface{}) {
+	l.print(ctx, logrus.WarnLevel, args...)
+}
+
+// WarnWithContext logs a message and context (tarceid...) at level Warn.
+func (l *daprLogger) WarnWithContextf(ctx context.Context, format string, args ...interface{}) {
+	l.printf(ctx, logrus.WarnLevel, format, args...)
 }
 
 // Error logs a message at level Error.
 func (l *daprLogger) Error(args ...interface{}) {
-	l.logger.Log(logrus.ErrorLevel, args...)
+	l.print(context.Background(), logrus.ErrorLevel, args...)
 }
 
 // Errorf logs a message at level Error.
 func (l *daprLogger) Errorf(format string, args ...interface{}) {
-	l.logger.Logf(logrus.ErrorLevel, format, args...)
+	l.printf(context.Background(), logrus.ErrorLevel, format, args...)
+}
+
+// ErrorWithContext logs a message and context (traceid...) at level Error.
+func (l *daprLogger) ErrorWithContext(ctx context.Context, args ...interface{}) {
+	l.print(ctx, logrus.ErrorLevel, args...)
+}
+
+// ErrorWithContext logs a message and context (traceid...) at level Error.
+func (l *daprLogger) ErrorWithContextf(ctx context.Context, format string, args ...interface{}) {
+	l.printf(ctx, logrus.ErrorLevel, format, args...)
 }
 
 // Fatal logs a message at level Fatal then the process will exit with status set to 1.
 func (l *daprLogger) Fatal(args ...interface{}) {
-	l.logger.Fatal(args...)
+	l.print(context.Background(), logrus.FatalLevel, args...)
 }
 
 // Fatalf logs a message at level Fatal then the process will exit with status set to 1.
 func (l *daprLogger) Fatalf(format string, args ...interface{}) {
-	l.logger.Fatalf(format, args...)
+	l.printf(context.Background(), logrus.FatalLevel, format, args...)
+}
+
+// FatalWithContext logs a message and context (traceid...) at level Fatal then the process will exit with status set to 1.
+func (l *daprLogger) FatalWithContext(ctx context.Context, args ...interface{}) {
```

Comment:
Rather than creating a large number of `*level*WithContext` methods, and then alter each method to pass a context, please create a method `func (l *daprLogger) WithContext(ctx context. Context) Logger` which returns a new Logger with the ID field set. For example:

```go
func (l *daprLogger) WithContext(ctx context. Context) Logger {
	return l.WithFields(...)
}
```